pkg/apis/hive/v1/agent/platform.go

```diff
@@ -0,0 +1,136 @@
+	// NetworkType is the type of network to install. The default is OpenShiftSDN
+	//
+	// +kubebuilder:default=OpenShiftSDN
+	// +optional
+	//NetworkType string `json:"networkType,omitempty"`
+
+	// MachineNetwork is the list of IP address pools for machines.
+	// This field replaces MachineCIDR, and if set MachineCIDR must
+	// be empty or match the first entry in the list.
+	// Default is 10.0.0.0/16 for all platforms other than libvirt.
+	// For libvirt, the default is 192.168.126.0/24.
+	//
+	// +optional
+	MachineNetwork []MachineNetworkEntry `json:"machineNetwork,omitempty"`
+
+	// ClusterNetwork is the list of IP address pools for pods.
+	// Default is 10.128.0.0/14 and a host prefix of /23.
+	//
+	// +optional
+	ClusterNetwork []ClusterNetworkEntry `json:"clusterNetwork,omitempty"`
+
+	// ServiceNetwork is the list of IP address pools for services.
+	// Default is 172.30.0.0/16.
+	// NOTE: currently only one entry is supported.
+	//
+	// +kubebuilder:validation:MaxItems=1
+	// +optional
+	ServiceNetwork []IPNet `json:"serviceNetwork,omitempty"`
+}
+
+// MachineNetworkEntry is a single IP address block for node IP blocks.
+type MachineNetworkEntry struct {
+	// CIDR is the IP block address pool for machines within the cluster.
+	CIDR IPNet `json:"cidr"`
+}
+
+// ClusterNetworkEntry is a single IP address block for pod IP blocks. IP blocks
+// are allocated with size 2^HostSubnetLength.
+type ClusterNetworkEntry struct {
+	// CIDR is the IP block address pool.
+	CIDR IPNet `json:"cidr"`
+
+	// HostPrefix is the prefix size to allocate to each node from the CIDR.
+	// For example, 24 would allocate 2^8=256 adresses to each node. If this
+	// field is not used by the plugin, it can be left unset.
+	// +optional
+	HostPrefix int32 `json:"hostPrefix,omitempty"`
+}
+
+// Proxy defines the proxy settings for the cluster.
+// At least one of HTTPProxy or HTTPSProxy is required.
+type Proxy struct {
+	// HTTPProxy is the URL of the proxy for HTTP requests.
+	// +optional
+	HTTPProxy string `json:"httpProxy,omitempty"`
+
+	// HTTPSProxy is the URL of the proxy for HTTPS requests.
+	// +optional
+	HTTPSProxy string `json:"httpsProxy,omitempty"`
+
+	// NoProxy is a comma-separated list of domains and CIDRs for which the proxy should not be used.
+	// +optional
+	NoProxy string `json:"noProxy,omitempty"`
+}
+
+type ClusterProgressInfo struct {
+	// progress info
+	ProgressInfo string `json:"progressInfo,omitempty"`
+	// Time at which the cluster install progress was last updated.
+	LastProgressUpdateTime *metav1.Time `json:"lastProgressUpdateTime,omitempty"`
+}
+
+type HostNetwork struct {
+	// cidr
+	Cidr string `json:"cidr,omitempty"`
+	// host ids
+	HostIds []string `json:"hostIDs"`
+}
+
+// ClusterStatus defines the observed state of Cluster
+type ClusterStatus struct {
+	State                        string              `json:"state,omitempty"`
+	StateInfo                    string              `json:"stateInfo,omitempty"`
```

Comment:
i think we should replace state info with some conditions. the state enum is good enough to convey what state we are in, and any other info should be in conditions.

Reply:
StateInfo provides error messages that aren't all formalized

Reply:
Conditions are typically composed of a Type, a Reason, and a Message. An unformalized message is still quite likely ok there. I would agree with @abhinavdahiya, we really should get these boiled down into a handful of conditions for better compat with Kube APIs.

Reply:
I agree with choosing the former unless we have a very good reason otherwise.

Reply:
I will try to sketch something out in the next couple days.

Reply:
Interesting tidbit from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md

*"Some resources in the v1 API contain fields called phase, and associated message, reason, and other status fields. The pattern of using phase is deprecated. Newer API types should use conditions instead. Phase was essentially a state-machine enumeration field, that contradicted system-design principles and hampered evolution, since adding new enum values breaks backward compatibility. Rather than encouraging clients to infer implicit properties from phases, we prefer to explicitly expose the individual conditions that clients need to monitor. Conditions also have the benefit that it is possible to create some conditions with uniform meaning across all resource types, while still exposing others that are unique to specific resource types. See #7856 for more details and discussion."*

Based on that, I think we drop the State enum as well.

@avishayt @filanov could one of you add the list of valid states here so we can discuss how to model.

Reply:
status:
        type: string
        description: Status of the OpenShift cluster.
        enum:
          - insufficient
          - ready
          - error
          - preparing-for-installation
          - pending-for-input
          - installing
          - finalizing
          - installed
          - adding-hosts
          - cancelled
          - installing-pending-user-action

Reply:
We will add one condition per each of these states. Assisted controllers will translate their internal state machine into a set of these conditions.